Modules/posixmodule.c

```
@@ -4530,6 +4530,86 @@ os_listmounts_impl(PyObject *module, path_t *volume)
 }
 
 
+/*[clinic input]
+os._path_isdevdrive
+
+    path: path_t
+
+Determines whether the specified path is on a Windows Dev Drive.
+
+[clinic start generated code]*/
+
+static PyObject *
+os__path_isdevdrive_impl(PyObject *module, path_t *path)
+/*[clinic end generated code: output=1f437ea6677433a2 input=ee83e4996a48e23d]*/
+{
+#ifndef PERSISTENT_VOLUME_STATE_DEV_VOLUME
+    const int PERSISTENT_VOLUME_STATE_DEV_VOLUME = 0x00002000;
+#endif
+    int err = 0;
+    PyObject *r = NULL;
+    wchar_t volume[MAX_PATH];
```

I personally wouldn't allocate such a large thing on the stack.  It's also not critical and can be improved to malloc/free later if it ever bothers anyone.

---

I'm surprised the path of the volume is being limited to a hard-coded length of `MAX_PATH`. The dev drive could be mounted on a long path.

If 259 characters isn't enough for the path (i.e. `ERROR_FILENAME_EXCED_RANGE`), or the call 'succeeds' without storing a trailing backslash (i.e. the buffer is actually 1 character too small), then I'd prefer to loop, doubling the size of a heap allocated buffer on each pass.

Note that the recommendation in the WinAPI docs to first get an absolute path via `GetFullPathNameW()` is nonsense. The length of the input path generally has nothing to do with the length of the path that's returned by `GetVolumePathNameW()`. For example, here's a volume that's mounted on a long path, but accessed using a symlink or junction that's a short path:

```pycon
>>> win32file.GetVolumePathName(r'C:\Temp\test\volume', 520)
'C:\\Temp\\test\\longpath\\aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\\volume\\'
```

---

> If 259 characters isn't enough for the path (i.e. `ERROR_FILENAME_EXCED_RANGE`), or the call 'succeeds' without storing a trailing backslash (i.e. the buffer is actually 1 character too small), then I'd prefer to loop, doubling the size of a heap allocated buffer on each pass.

Yeah, this is the reliable way. But it's incredibly unlikely for a volume path to be so long, and if it is then so much other stuff is going to break already that users would change it (e.g. the C++ compiler still does not support extra long paths). If we had a better metaphor in C for this kind of loop, I'd just do it, but it'll double the length of this function for almost no benefit.

We won't fail hard in this case, it'll just misidentify a Dev Drive as a regularly formatted drive.